# The constant capacitance procedure of Goldberg (1985), applied to Li⁺ and Co²⁺ on dolomite

**A step-for-step transcription of the method, with this system's sites, reactions and parameters.**

Goldberg, S. (1985). *Chemical modeling of anion competition on goethite using the constant
capacitance model.* Soil Science Society of America Journal, 49(4), 851–856.

This notebook follows that paper in its own order, with its own equation numbering in square
brackets, substituting the dolomite surface for goethite and Li⁺ and Co²⁺ for phosphate, selenite
and silicate. It is self-contained: run it top to bottom, nothing is read from a file.

Goldberg's structure is also this project's structure. He fits surface complexation constants on
**single-adsorbate** systems with the acid–base constants and the capacitance held fixed, then uses
those constants, with **no parameter fitted at all**, to predict a **competitive** two-adsorbate
system. That is the path from the single-ion Li and Co batches to produced water.

## 1. The four assumptions

Goldberg states four. Three carry over unchanged; one does not, and where a mimic breaks is worth
more than where it holds.

| | Goldberg (1985), goethite | This system, dolomite |
|---|---|---|
| (i) | adsorption proceeds by **ligand exchange** | adsorption proceeds by **proton exchange** at the carbonate site. The adsorbates are cations, so the exchanged ion is H⁺ rather than OH⁻. The formalism is identical; only the reaction differs |
| (ii) | all surface complexes are **inner-sphere** | assumed here too, and it is this assumption that Zachara et al. (1991) and Brady et al. (1999) put in doubt for hydrated divalent cations in concentrated NaCl |
| (iii) | **no complexes with the background electrolyte** | **violated, and it cannot be waived.** The background is 0.684 M NaCl, in which CoCl⁺ carries about 20 % of the dissolved cobalt. An aqueous speciation stage must precede the surface calculation |
| (iv) | **linear** σ–ψ relation | assumed here too, Eq. [1] |

## 2. Charge and potential [Goldberg Eq. 1]

Goldberg writes the constant capacitance relation on a **volume-of-suspension** basis:

$$\sigma = \frac{C\,S\,a}{F}\,\psi \qquad [1]$$

with σ in mol$_c$ m⁻³, $C$ in F m⁻², $S$ in m² kg⁻¹, $a$ in kg m⁻³, $F$ in C mol⁻¹ and ψ in V.
This is the same statement as the area-based $\sigma_{area} = C\psi$, multiplied by $S a/F$ to put
charge on a per-volume basis. Both are printed below, because confusing the two — and confusing
either with a total-ion mass balance — is the origin of the recurring impression that the computed
surface charge is too large.

## 3. Surface functional groups and their acid–base reactions [Goldberg Eqs. 2–3]

Goethite has one amphoteric group, FeOH. Dolomite has **three** primary hydration sites in 1:1:2
stoichiometry (Pokrovsky et al., 1999), and only one binds metal cations, so Goldberg's pair becomes
five reactions:

$$\begin{aligned}
\mathrm{>CaOH^0 + H^+} &\rightleftharpoons \mathrm{>CaOH_2^+} & [2a]\\
\mathrm{>MgOH^0 + H^+} &\rightleftharpoons \mathrm{>MgOH_2^+} & [2b]\\
\mathrm{>CaOH^0} &\rightleftharpoons \mathrm{>CaO^- + H^+} & [3a]\\
\mathrm{>MgOH^0} &\rightleftharpoons \mathrm{>MgO^- + H^+} & [3b]\\
\mathrm{>CO_3H^0} &\rightleftharpoons \mathrm{>CO_3^- + H^+} & [3c]
\end{aligned}$$

[3c] has no counterpart in the goethite system: it is the carbonate site, and the deprotonated
>CO₃⁻ is what binds the cation.

## 4. Adsorbate surface complexation reactions [Goldberg Eqs. 4–7]

Goldberg's [4]–[7] exchange an anion for a surface hydroxyl. The cation analogue exchanges a proton
at the carbonate site. The two constants this work determines:

$$\begin{aligned}
\mathrm{>CO_3H^0 + Li^+} &\rightleftharpoons \mathrm{>CO_3Li^0 + H^+} & [4]\\
\mathrm{>CO_3H^0 + Co^{2+}} &\rightleftharpoons \mathrm{>CO_3Co^+ + H^+} & [5]
\end{aligned}$$

Two further reactions of identical form are **not** adjustable, because the dolomite supplies their
cations by dissolving. Goldberg has no equivalent, because goethite does not dissolve:

$$\begin{aligned}
\mathrm{>CO_3H^0 + Ca^{2+}} &\rightleftharpoons \mathrm{>CO_3Ca^+ + H^+} & [6]\\
\mathrm{>CO_3H^0 + Mg^{2+}} &\rightleftharpoons \mathrm{>CO_3Mg^+ + H^+} & [7]
\end{aligned}$$

and the carbonate complexes of the hydroxyl sites, which bind no metal but do carry charge:

$$\begin{aligned}
\mathrm{>MeOH^0 + CO_3^{2-} + 2H^+} &\rightleftharpoons \mathrm{>MeHCO_3^0 + H_2O} & [8]\\
\mathrm{>MeOH^0 + CO_3^{2-} + H^+} &\rightleftharpoons \mathrm{>MeCO_3^- + H_2O} & [9]
\end{aligned}$$

## 5. Intrinsic conditional equilibrium constants [Goldberg Eqs. 8–13]

With $\{\}$ for aqueous activity and $[\,]$ for surface concentration (mol L⁻¹):

$$K_{Li} = \frac{[\mathrm{>CO_3Li^0}]\{H^+\}}{[\mathrm{>CO_3H^0}]\{Li^+\}}\;[10]
\qquad K_{Co} = \frac{[\mathrm{>CO_3Co^+}]\{H^+\}}{[\mathrm{>CO_3H^0}]\{Co^{2+}\}}\;[11]$$

$$K_{Ca} = \frac{[\mathrm{>CO_3Ca^+}]\{H^+\}}{[\mathrm{>CO_3H^0}]\{Ca^{2+}\}}\;[12]
\qquad K_{Mg} = \frac{[\mathrm{>CO_3Mg^+}]\{H^+\}}{[\mathrm{>CO_3H^0}]\{Mg^{2+}\}}\;[13]$$

## 6. Mass balance on the surface functional groups [Goldberg Eq. 14]

Goldberg needs one site balance; dolomite needs three:

$$\begin{aligned}
[\mathrm{>CO_3}]_T &= [\mathrm{>CO_3H^0}]+[\mathrm{>CO_3^-}]+[\mathrm{>CO_3Li^0}]+[\mathrm{>CO_3Co^+}]+[\mathrm{>CO_3Ca^+}]+[\mathrm{>CO_3Mg^+}] & [14a]\\
[\mathrm{>CaOH}]_T &= [\mathrm{>CaOH^0}]+[\mathrm{>CaOH_2^+}]+[\mathrm{>CaO^-}]+[\mathrm{>CaHCO_3^0}]+[\mathrm{>CaCO_3^-}] & [14b]\\
[\mathrm{>MgOH}]_T &= [\mathrm{>MgOH^0}]+[\mathrm{>MgOH_2^+}]+[\mathrm{>MgO^-}]+[\mathrm{>MgHCO_3^0}]+[\mathrm{>MgCO_3^-}] & [14c]
\end{aligned}$$

## 7. Charge balance [Goldberg Eq. 15]

$$\sigma = [\mathrm{>CaOH_2^+}]+[\mathrm{>MgOH_2^+}]+[\mathrm{>CO_3Co^+}]+[\mathrm{>CO_3Ca^+}]+[\mathrm{>CO_3Mg^+}]
-[\mathrm{>CO_3^-}]-[\mathrm{>CaO^-}]-[\mathrm{>MgO^-}]-[\mathrm{>CaCO_3^-}]-[\mathrm{>MgCO_3^-}] \quad [15]$$

>CO₃Li⁰, >CaHCO₃⁰ and >MgHCO₃⁰ are neutral and contribute nothing. **Dissolved Ca²⁺ and Mg²⁺ from
dissolution appear nowhere in Eq. [15].** A surface charge computed from a total-ion mass balance on
the solution is a different quantity: it carries the dissolution signal and at acidic pH exceeds the
whole site capacity many times over.

## 8. Intrinsic versus conditional constants [Goldberg Eqs. 16–17]

$$K(\mathrm{int}) = {}^cK\,\exp\!\left(\frac{\Delta Z\,F\,\psi}{RT}\right) \quad [16]
\qquad {}^cK = \text{the same quotient in concentrations} \quad [17]$$

ΔZ is the change in the charge of the surface species, and it decides which constants feel the
electrostatics at all:

| Reaction | charge before → after | ΔZ | factor |
|---|---|---|---|
| [4] lithium | 0 → 0 | **0** | **none** |
| [5] cobalt | 0 → +1 | +1 | exp(+Fψ/RT) |
| [6], [7] Ca, Mg | 0 → +1 | +1 | exp(+Fψ/RT) |
| [2a], [2b] | 0 → +1 | +1 | exp(+Fψ/RT) |
| [3a]–[3c] | 0 → −1 | −1 | exp(−Fψ/RT) |
| [8] | 0 → 0 | 0 | none |
| [9] | 0 → −1 | −1 | exp(−Fψ/RT) |

Because ΔZ = 0 for lithium, the lithium constant is **electrostatically blind**: its CCM and
non-electrostatic values are identical to rounding. That is a property of the stoichiometry, not
evidence that the constant is well constrained.

## 9. The model, in code

Everything below is copied verbatim from the project's main notebook, so this document cannot drift from the model it describes.

### 9.1 Imports and the fixed parameter block

In [1]:
import numpy as np, pandas as pd
from scipy.optimize import brentq
pd.set_option('display.width', 130); pd.set_option('display.precision', 4)

In [2]:
F, R, T = 96485.0, 8.314, 298.15
FRT = F/(R*T)                                    # 38.92 V^-1

# --- experimental set-up (spreadsheet "Set up") ---
DOLOMITE_GL = 60.0          # 6 g per 100 mL
V_L, M_G   = 0.100, 6.0     # litres of solution, grams of dolomite per bottle
SSA        = 0.76           # m2/g (BET; 0.76 to 0.84 reported)
S_AREA     = DOLOMITE_GL*SSA                     # 45.6 m2 per litre
I_BATCH    = 0.684          # mol/L, 40 g/L NaCl
NACL_M     = 40.0/58.44
MM = dict(Li=6.941, Co=58.933, Ca=40.078, Mg=24.305)

# --- site density: Pokrovsky 1999, dolomite, 1:1:2 Ca:Mg:CO3 ---
SITE_DENS = dict(CO3=14.0e-6, Ca=7.0e-6, Mg=7.0e-6)      # mol/m2
S_T = {k: v*S_AREA for k, v in SITE_DENS.items()}         # mol/L of suspension

# --- capacitance: carbonate formula C = sqrt(I)/alpha ---
# alpha = 0.004 C mol^-1/2 V^-1 m^3 is the value Pokrovsky, Schott & Thomas (1999) fitted for
# dolomite (their section 3.3), and it is used here because that is the paper this work follows.
# A value of 0.006 is sometimes quoted from later carbonate work; the choice is immaterial to the
# result - see the alpha sensitivity printed with the fits - so the source paper's value is kept.
ALPHA = 0.004
C_CAP = np.sqrt(I_BATCH)/ALPHA                            # F/m2

# --- surface constants: Pokrovsky, Schott & Thomas (1999) Table 3, dolomite columns ---
# Transcribed from the page. Reaction : calcite (Ca) | magnesite (Mg) | dolomite Ca | dolomite Mg
#  1. >CO3H0 = >CO3- + H+            : -5.1  | -4.65 +/- 0.15 | -4.8 +/- 0.2 | -4.8 +/- 0.2
#  2. >CO3H0 + Me2+ = >CO3Me+ + H+   : -1.7  | -2.2  +/- 0.15 | -1.8 +/- 0.2 | -2.0 +/- 0.2
#  3. >MeOH0 = >MeO- + H+            : -12   | -12   +/- 1    | -12  +/- 2   | -12  +/- 2
#  4. >MeOH0 + H+ = >MeOH2+          : 11.5  | 10.6  +/- 0.15 | 11.5 +/- 0.2 | 10.6 +/- 0.2
#  5. >MeOH0 + CO3-2 + 2H+ = >MeHCO3 : -3.5  | -2.4  +/- 0.5  | -4.0 +/- 0.5 | -3.5 +/- 0.5
#  6. >MeOH0 + CO3-2 + H+ = >MeCO3-  : 17.1  | 14.4  +/- 0.15 | 16.6 +/- 0.2 | 15.4 +/- 0.2
# Re-checked line by line against Table 3 of the PDF: all eleven constants and all eleven
# uncertainties above match the two dolomite columns exactly.
# Site density, same paper, section 3.3: "The site density was assumed to be 7 mmol/m2 for Ca and
# Mg and 14 mmol/m2 for carbonate. This value corresponds to eight metal or carbonate sites per
# square nanometer" (1:1:2 stoichiometry). The printed UNIT is wrong by 10^3 - 14 mmol/m2 would be
# 8400 sites/nm2, not eight - and the code uses umol/m2, which reproduces the paper's own
# eight-sites-per-nm2 statement (14e-6 mol/m2 x 6.022e23 = 8.4 sites/nm2). This is the same class
# of slip as the m2/mL-for-m2/L in the manuscript's methods, noted in the capacity check.
LOGK_UNC = {"CO3_deprot":0.2,"CO3Ca":0.2,"CO3Mg":0.2,"CaOH2":0.2,"CaO":2.0,"CaHCO3":0.5,
            "CaCO3":0.2,"MgOH2":0.2,"MgO":2.0,"MgHCO3":0.5,"MgCO3":0.2}
LOGK_SURF = {
    "CO3_deprot": -4.8,   # >CO3H0 = >CO3- + H+
    "CO3Ca":      -1.8,   # >CO3H0 + Ca2+ = >CO3Ca+ + H+      (competitor)
    "CO3Mg":      -2.0,   # >CO3H0 + Mg2+ = >CO3Mg+ + H+      (competitor)
    "CaOH2":      11.5,   # >CaOH0 + H+ = >CaOH2+             (carries charge, no metal)
    "CaO":       -12.0,   # >CaOH0 = >CaO- + H+
    "CaHCO3":     -4.0,   # >CaOH0 + CO3-2 + 2H+ = >CaHCO3(0) + H2O
    "CaCO3":      16.6,   # >CaOH0 + CO3-2 + H+  = >CaCO3- + H2O
    "MgOH2":      10.6,
    "MgO":       -12.0,
    "MgHCO3":     -3.5,
    "MgCO3":      15.4,
}
# --- open-system carbonate ---
PCO2, LOGKH = 10**-3.5, -1.469
A_CO2 = 10**LOGKH*PCO2
LK_H2CO3, LK_HCO3 = -6.345, 10.329            # CO2+H2O=HCO3-+H+ ; CO3-2+H+=HCO3-

# --- aqueous complex FORMATION log beta from free ions (25 C, I = 0; NIST/MINTEQ values) ---
# entries: (name, log beta, {component: stoich}, charge)   components: Me, Ca, Mg, Na, Cl, CO3, H
AQ = {
  "Ca": [("CaCl+",-0.70,{"Ca":1,"Cl":1},1),("CaCO3",3.22,{"Ca":1,"CO3":1},0),
         ("CaHCO3+",11.43,{"Ca":1,"CO3":1,"H":1},1),("CaOH+",-12.70,{"Ca":1,"H":-1},1)],
  "Mg": [("MgCl+",-0.14,{"Mg":1,"Cl":1},1),("MgCO3",2.98,{"Mg":1,"CO3":1},0),
         ("MgHCO3+",11.40,{"Mg":1,"CO3":1,"H":1},1),("MgOH+",-11.79,{"Mg":1,"H":-1},1)],
  "Na": [("NaCl",-0.78,{"Na":1,"Cl":1},0),("NaCO3-",1.27,{"Na":1,"CO3":1},-1),
         ("NaHCO3",10.48,{"Na":1,"CO3":1,"H":1},0),("NaOH",-14.18,{"Na":1,"H":-1},0)],
  "Co": [("CoCl+",0.30,{"Co":1,"Cl":1},1),("CoCl2",-0.20,{"Co":1,"Cl":2},0),
         ("CoCO3",4.23,{"Co":1,"CO3":1},0),("CoHCO3+",12.20,{"Co":1,"CO3":1,"H":1},1),
         ("CoOH+",-9.65,{"Co":1,"H":-1},1),("Co(OH)2",-18.80,{"Co":1,"H":-2},0)],
  "Li": [("LiCl",-0.50,{"Li":1,"Cl":1},0),("LiOH",-13.64,{"Li":1,"H":-1},0),
         ("LiCO3-",0.90,{"Li":1,"CO3":1},-1)],
}
Z_ION = dict(Li=1, Co=2, Ca=2, Mg=2, Na=1)
# --- solubility products (log Ksp, 25 C) for the saturation check ---
LOGKSP = {"CoCO3 (sphaerocobaltite)": -9.98, "Calcite": -8.48, "Dolomite (ordered)": -17.09,
          "Li2CO3": -2.5}   # Li2CO3 approximate; it is far undersaturated regardless

### 9.2 The measured data (single-ion batches, days 0–6)

In [3]:
raw = [  # metal, batch, day, Ca_ppm, Mg_ppm, Me_ppm, pH_measured
 ("Co","pH6",0, 0.100, 0.100, 80.513, 6.320),("Co","pH6",2, 0.100, 0.100, 77.905, 7.400),
 ("Co","pH6",4, 0.100, 8.143, 76.411, 7.477),("Co","pH6",6, 0.126,12.941, 74.393, 8.090),
 ("Co","pH2",0, 0.100, 0.100, 80.513, 2.350),("Co","pH2",2,57.477,17.461, 79.661, 4.123),
 ("Co","pH2",4,24.178,45.592, 79.795, 5.660),("Co","pH2",6,21.658,51.379, 80.398, 7.110),
 ("Li","pH6",0, 0.663, 0.100,138.430, 6.420),("Li","pH6",2,37.633, 0.100,134.657, 7.647),
 ("Li","pH6",4,54.820, 0.100,136.722, 7.923),("Li","pH6",6,49.892, 0.870,135.262, 8.257),
 ("Li","pH2",0, 0.663, 0.100,138.430, 2.250),("Li","pH2",2,39.190, 0.100,136.263, 4.257),
 ("Li","pH2",4,79.452,21.316,135.956, 5.727),("Li","pH2",6,87.738,20.547,135.592, 7.047),
]
data = pd.DataFrame(raw, columns=["metal","batch","day","Ca_ppm","Mg_ppm","Me_ppm","pH"])
data["C0_ppm"] = data.groupby(["metal","batch"])["Me_ppm"].transform("first")
data["removal_%"] = 100*(data.C0_ppm - data.Me_ppm)/data.C0_ppm
eq = data[data.day == 6].reset_index(drop=True)      # the four equilibrium points

print(data.to_string(index=False))
print("\nThe four equilibrium points (day 6). Note the MEASURED pH against the nominal label:")
print(eq[["metal","batch","pH","Ca_ppm","Mg_ppm","C0_ppm","Me_ppm","removal_%"]].to_string(index=False))

metal batch  day  Ca_ppm  Mg_ppm  Me_ppm    pH  C0_ppm  removal_%
   Co   pH6    0   0.100   0.100  80.513 6.320  80.513     0.0000
   Co   pH6    2   0.100   0.100  77.905 7.400  80.513     3.2392
   Co   pH6    4   0.100   8.143  76.411 7.477  80.513     5.0948
   Co   pH6    6   0.126  12.941  74.393 8.090  80.513     7.6013
   Co   pH2    0   0.100   0.100  80.513 2.350  80.513     0.0000
   Co   pH2    2  57.477  17.461  79.661 4.123  80.513     1.0582
   Co   pH2    4  24.178  45.592  79.795 5.660  80.513     0.8918
   Co   pH2    6  21.658  51.379  80.398 7.110  80.513     0.1428
   Li   pH6    0   0.663   0.100 138.430 6.420 138.430     0.0000
   Li   pH6    2  37.633   0.100 134.657 7.647 138.430     2.7256
   Li   pH6    4  54.820   0.100 136.722 7.923 138.430     1.2338
   Li   pH6    6  49.892   0.870 135.262 8.257 138.430     2.2885
   Li   pH2    0   0.663   0.100 138.430 2.250 138.430     0.0000
   Li   pH2    2  39.190   0.100 136.263 4.257 138.430     1.5654
   Li   pH

### 9.3 Aqueous speciation — the stage Goldberg's assumption (iii) does not need

In [4]:
def davies(I):
    f = -0.509*(np.sqrt(I)/(1+np.sqrt(I)) - 0.3*I)
    return {0:1.0, 1:10**f, 2:10**(4*f)}

def speciate(pH, tot, metal):
    """tot: dict of totals mol/L for Ca, Mg, Na, Cl and the metal (key = metal name)."""
    aH = 10**-pH
    aHCO3 = 10**LK_H2CO3*A_CO2/aH
    aCO3  = aHCO3/(10**LK_HCO3*aH)
    cations = ["Ca","Mg","Na",metal]
    I = 0.5*(tot["Na"]+tot["Cl"]+sum(Z_ION[c]**2*tot[c] for c in cations if c!="Na"))
    a = {c: tot[c]*0.5 for c in cations}; a["Cl"] = tot["Cl"]*0.7
    fixed = {"CO3":aCO3,"H":aH}
    for _ in range(300):
        g = davies(I)
        def conc(name, lb, st, z, act):
            v = 10**lb
            for comp, n in st.items():
                x = act[comp] if comp in act else fixed[comp]
                v *= x**n
            return v/g[abs(z)]
        # cations
        for c in cations:
            D = 1/g[Z_ION[c]]
            for name, lb, st, z in AQ[c]:
                D += conc(name, lb, {k:v for k,v in st.items() if k!=c}, z, {**a, **fixed})
            a[c] = tot[c]/D
        # chloride
        Dcl = 1/g[1]
        for c in cations:
            for name, lb, st, z in AQ[c]:
                if "Cl" in st:
                    Dcl += st["Cl"]*conc(name, lb, {k:v for k,v in st.items() if k!="Cl"}, z, {**a, **fixed})
        aCl_new = tot["Cl"]/Dcl
        # ionic strength
        species = []
        for c in cations:
            for name, lb, st, z in AQ[c]:
                species.append((z, conc(name, lb, st, z, {**a, "Cl":aCl_new, **fixed})))
        Inew = 0.5*(sum(Z_ION[c]**2*a[c]/g[Z_ION[c]] for c in cations) + aCl_new/g[1]
                    + aH/g[1] + 1e-14/aH/g[1] + aHCO3/g[1] + 4*aCO3/g[2]
                    + sum(z*z*cc for z, cc in species))
        done = abs(aCl_new - a["Cl"]) < 1e-12 and abs(Inew - I) < 1e-9
        a["Cl"] = aCl_new; I = 0.5*I + 0.5*Inew
        if done: break
    g = davies(I)
    # metal distribution
    free = a[metal]/g[Z_ION[metal]]
    dist = {metal+"2+" if Z_ION[metal]==2 else metal+"+": free}
    for name, lb, st, z in AQ[metal]:
        dist[name] = conc(name, lb, st, z, {**a, **fixed})
    frac_free = free/tot[metal]
    SI = {"CoCO3 (sphaerocobaltite)": np.log10(a["Co"]*aCO3) - LOGKSP["CoCO3 (sphaerocobaltite)"] if metal=="Co" else np.nan,
          "Calcite": np.log10(a["Ca"]*aCO3) - LOGKSP["Calcite"],
          "Dolomite (ordered)": np.log10(a["Ca"]*a["Mg"]*aCO3**2) - LOGKSP["Dolomite (ordered)"],
          "Li2CO3": np.log10(a["Li"]**2*aCO3) - LOGKSP["Li2CO3"] if metal=="Li" else np.nan}
    return dict(I=I, aH=aH, aCO3=aCO3, aHCO3=aHCO3, aCa=a["Ca"], aMg=a["Mg"], aMe=a[metal],
                aCl=a["Cl"], g=g, frac_free=frac_free, dist=dist, SI=SI)

def totals_for(row):
    m = row.metal
    return {"Ca":row.Ca_ppm/MM["Ca"]/1e3, "Mg":row.Mg_ppm/MM["Mg"]/1e3, "Na":NACL_M,
            "Cl":NACL_M, m:row.Me_ppm/MM[m]/1e3}

_sp = pd.DataFrame([dict(metal=r.metal, batch=r.batch, day=r.day, pH=r.pH,
                         I=(_s := speciate(r.pH, totals_for(r), r.metal))["I"],
                         a_Me=_s["aMe"], frac_free=_s["frac_free"],
                         a_Ca=_s["aCa"], a_Mg=_s["aMg"], a_CO3=_s["aCO3"])
                    for r in data[data.day > 0].itertuples()])
print(_sp.to_string(index=False, float_format=lambda x: f"{x:.3g}"))
print("\nThis is the stage Goldberg's assumption (iii) lets him skip. At 0.684 M NaCl it cannot be")
print("skipped: roughly a fifth of the dissolved cobalt is CoCl+, and the surface reaction draws")
print("only on the FREE ion.")

metal batch  day   pH     I     a_Me  frac_free     a_Ca     a_Mg    a_CO3
   Co   pH6    2  7.4 0.648 0.000281      0.691 7.37e-07 1.11e-06 1.44e-07
   Co   pH6    4 7.48 0.648 0.000275       0.69 7.37e-07    9e-05 2.05e-07
   Co   pH6    6 8.09 0.649 0.000261      0.671 9.26e-07 0.000143 3.44e-06
   Co   pH2    2 4.12 0.652 0.000289      0.694 0.000425 0.000193 4.01e-14
   Co   pH2    4 5.66 0.652 0.000289      0.694 0.000179 0.000505 4.75e-11
   Co   pH2    6 7.11 0.653 0.000291      0.692  0.00016 0.000569 3.78e-08
   Li   pH6    2 7.65 0.656    0.013      0.898 0.000278 1.11e-06 4.48e-07
   Li   pH6    4 7.92 0.657   0.0132      0.898 0.000405 1.11e-06  1.6e-06
   Li   pH6    6 8.26 0.657    0.013      0.898 0.000367 9.61e-06 7.43e-06
   Li   pH2    2 4.26 0.656   0.0131      0.898  0.00029 1.11e-06 7.43e-14
   Li   pH2    4 5.73 0.659   0.0131      0.898  0.00059 0.000237 6.47e-11
   Li   pH2    6 7.05  0.66   0.0131      0.898 0.000651 0.000228 2.82e-08

This is the stage Goldbe

### 9.4 The site balances [14], the charge balance [15] and the CCM loop [1]

In [5]:
def surface(s, metal, Gamma_known=None, logK_Me=None, electrostatic=True, C_cap=None):
    """Solve the three site balances and the CCM loop at one data point.
    Either the metal coverage is KNOWN (point-by-point mode, Gamma_known in mol/m2) or a
    trial log K is given (forward mode). Returns all surface concentrations (mol/L), psi, sigma."""
    aH, aCO3, aCa, aMg, aMe = s["aH"], s["aCO3"], s["aCa"], s["aMg"], s["aMe"]
    z = Z_ION[metal]; dZ = z - 1                      # charge of >CO3Me
    def state(psi):
        b = np.exp(-FRT*psi)                          # a species of charge q carries b**q
        rm  = 10**LOGK_SURF["CO3_deprot"]/aH/b        # >CO3-   / >CO3H0
        rCa = 10**LOGK_SURF["CO3Ca"]*aCa/aH*b         # >CO3Ca+ / >CO3H0
        rMg = 10**LOGK_SURF["CO3Mg"]*aMg/aH*b
        if Gamma_known is not None:
            cMe = Gamma_known*S_AREA
            cH0 = (S_T["CO3"] - cMe)/(1 + rm + rCa + rMg)
        else:
            rMe = 10**logK_Me*aMe/aH*b**dZ
            cH0 = S_T["CO3"]/(1 + rm + rCa + rMg + rMe)
            cMe = rMe*cH0
        cm, cCa, cMg = rm*cH0, rCa*cH0, rMg*cH0
        # hydroxyl sites (no metal)
        dCa = 1 + 10**LOGK_SURF["CaOH2"]*aH*b + 10**LOGK_SURF["CaO"]/aH/b \
                + 10**LOGK_SURF["CaHCO3"]*aCO3*aH**2 + 10**LOGK_SURF["CaCO3"]*aCO3*aH/b
        dMg = 1 + 10**LOGK_SURF["MgOH2"]*aH*b + 10**LOGK_SURF["MgO"]/aH/b \
                + 10**LOGK_SURF["MgHCO3"]*aCO3*aH**2 + 10**LOGK_SURF["MgCO3"]*aCO3*aH/b
        CaOH, MgOH = S_T["Ca"]/dCa, S_T["Mg"]/dMg
        sp = dict(CO3H0=cH0, CO3m=cm, CO3Ca=cCa, CO3Mg=cMg, CO3Me=cMe,
                  CaOH2=10**LOGK_SURF["CaOH2"]*aH*b*CaOH, CaO=10**LOGK_SURF["CaO"]/aH/b*CaOH,
                  CaCO3=10**LOGK_SURF["CaCO3"]*aCO3*aH/b*CaOH,
                  MgOH2=10**LOGK_SURF["MgOH2"]*aH*b*MgOH, MgO=10**LOGK_SURF["MgO"]/aH/b*MgOH,
                  MgCO3=10**LOGK_SURF["MgCO3"]*aCO3*aH/b*MgOH)
        sigma = F/S_AREA*(sp["CaOH2"]+sp["MgOH2"]+sp["CO3Ca"]+sp["CO3Mg"]+dZ*sp["CO3Me"]
                          - sp["CO3m"]-sp["CaO"]-sp["MgO"]-sp["CaCO3"]-sp["MgCO3"])
        return sp, sigma
    if electrostatic:
        C = C_CAP if C_cap is None else C_cap
        psi = brentq(lambda p: p - state(p)[1]/C, -0.5, 0.5, xtol=1e-7)        # 0.1 mV resolves to 1e-4 V
    else:
        psi = 0.0
    sp, sigma = state(psi)
    return dict(sp=sp, psi=psi, sigma=sigma, dZ=dZ)

## 10. Model parameters, in the layout of Goldberg's Table 1

In [6]:
S_KG   = SSA*1e3                       # m2 kg-1        (Goldberg's S)
A_KG   = DOLOMITE_GL                   # kg m-3         (Goldberg's a)
GOLD_K = C_CAP*S_KG*A_KG/F             # mol_c m-3 V-1  (the coefficient in Eq. [1])

print("Eq. [1]   sigma = (C.S.a/F).psi      sigma in mol_c m-3, psi in V")
print(f"   C  capacitance density        {C_CAP:8.1f} F m-2      (= sqrt(I)/alpha, alpha = {ALPHA})")
print(f"   S  specific surface area      {S_KG:8.1f} m2 kg-1    ({SSA} m2 g-1)")
print(f"   a  solid concentration        {A_KG:8.1f} kg m-3     ({DOLOMITE_GL} g L-1)")
print(f"   F  Faraday constant           {F:8.0f} C mol-1")
print(f"   -> C.S.a/F                    {GOLD_K:8.1f} mol_c m-3 V-1")

print("\nTable 1  Model parameters")
for nm, v, u in [("Capacitance density, C", C_CAP, "F m-2"),
                 ("Specific surface area, S", S_KG, "m2 kg-1"),
                 ("Solid concentration, a", A_KG, "kg m-3"),
                 ("Surface area per litre, S.a", S_AREA, "m2 L-1"),
                 ("Ionic strength, I", I_BATCH, "mol L-1"),
                 ("Max adsorption density [>CO3]_T", S_T["CO3"]*1e3, "mol m-3"),
                 ("   [>CaOH]_T", S_T["Ca"]*1e3, "mol m-3"),
                 ("   [>MgOH]_T", S_T["Mg"]*1e3, "mol m-3")]:
    print(f"  {nm:46} {v:12.4g}   {u}")

print("\nFixed intrinsic constants (Pokrovsky et al. 1999, Table 3, dolomite columns):")
NAMES = {"CO3_deprot":"[3c] >CO3H0 = >CO3- + H+",
         "CO3Ca":"[6]  >CO3H0 + Ca2+ = >CO3Ca+ + H+",
         "CO3Mg":"[7]  >CO3H0 + Mg2+ = >CO3Mg+ + H+",
         "CaOH2":"[2a] >CaOH0 + H+ = >CaOH2+", "MgOH2":"[2b] >MgOH0 + H+ = >MgOH2+",
         "CaO":"[3a] >CaOH0 = >CaO- + H+", "MgO":"[3b] >MgOH0 = >MgO- + H+",
         "CaHCO3":"[8]  >CaOH0 + CO3-2 + 2H+ = >CaHCO3(0) + H2O",
         "MgHCO3":"[8]  >MgOH0 + CO3-2 + 2H+ = >MgHCO3(0) + H2O",
         "CaCO3":"[9]  >CaOH0 + CO3-2 + H+ = >CaCO3- + H2O",
         "MgCO3":"[9]  >MgOH0 + CO3-2 + H+ = >MgCO3- + H2O"}
for k, lab in NAMES.items():
    print(f"   {lab:48} log K(int) = {LOGK_SURF[k]:+7.2f} +/- {LOGK_UNC[k]:.2f}")
print("\nAdjustable: [4] Li and [5] Co, one constant per metal (Goldberg fitted 2 to 3 per anion).")

Eq. [1]   sigma = (C.S.a/F).psi      sigma in mol_c m-3, psi in V
   C  capacitance density           206.8 F m-2      (= sqrt(I)/alpha, alpha = 0.004)
   S  specific surface area         760.0 m2 kg-1    (0.76 m2 g-1)
   a  solid concentration            60.0 kg m-3     (60.0 g L-1)
   F  Faraday constant              96485 C mol-1
   -> C.S.a/F                        97.7 mol_c m-3 V-1

Table 1  Model parameters
  Capacitance density, C                                206.8   F m-2
  Specific surface area, S                                760   m2 kg-1
  Solid concentration, a                                   60   kg m-3
  Surface area per litre, S.a                            45.6   m2 L-1
  Ionic strength, I                                     0.684   mol L-1
  Max adsorption density [>CO3]_T                      0.6384   mol m-3
     [>CaOH]_T                                         0.3192   mol m-3
     [>MgOH]_T                                         0.3192   mol m-3

Fixed int

## 11. Equations [1], [15] and [16] evaluated at the four equilibrium points

Both charge conventions are printed side by side. Their ratio is exactly $S a/F$; neither is the
total-ion mass balance.

In [7]:
LOGK_ME = {"Li": -2.33, "Co": -2.20}      # from the main notebook's fit; see it for how to read them

print(f"  {'point':18} {'psi (mV)':>9} {'sigma [C m-2]':>14} {'sigma [mol_c m-3]':>18} {'exp(F.psi/RT)':>14}")
for r in eq.itertuples():
    s_  = speciate(r.pH, totals_for(r), r.metal)
    st  = surface(s_, r.metal, logK_Me=LOGK_ME[r.metal])
    psi = st["psi"]
    tag = f"{r.metal} {r.batch} pH {r.pH:.2f}"
    print(f"  {tag:18} {1e3*psi:9.2f} {C_CAP*psi:14.4f} {GOLD_K*psi:18.4f} {np.exp(FRT*psi):14.4f}")

print("\nThe Boltzmann factor of Eq. [16] multiplies every dZ = +1 constant and divides every")
print("dZ = -1 constant. It leaves lithium untouched, because dZ = 0 for reaction [4].")

  point               psi (mV)  sigma [C m-2]  sigma [mol_c m-3]  exp(F.psi/RT)
  Co pH6 pH 8.09         -0.77        -0.1590            -0.0752         0.9705
  Co pH2 pH 7.11          4.07         0.8424             0.3981         1.1718
  Li pH6 pH 8.26          0.39         0.0799             0.0377         1.0151
  Li pH2 pH 7.05          5.78         1.1944             0.5645         1.2521

The Boltzmann factor of Eq. [16] multiplies every dZ = +1 constant and divides every
dZ = -1 constant. It leaves lithium untouched, because dZ = 0 for reaction [4].


## 12. Where this mimic is exact, and where it is not

**Exact.** Assumptions (i), (ii), (iv); Eqs. [1], [14], [15], [16], [17]; acid–base constants and
capacitance held fixed with only the adsorbate constants adjustable; a single-adsorbate fit followed
by a competitive prediction in which no parameter is refitted.

**Not exact**, each for a stated reason:

1. **Background electrolyte.** Assumption (iii) fails at 0.684 M NaCl, so §9.3 is inserted ahead of
   the surface calculation. Goldberg has no equivalent stage.
2. **The solid dissolves.** Goethite does not; dolomite does, supplying the competitors of [6] and
   [7], whose surface fractions therefore move point by point with the measured Ca and Mg.
3. **Capacitance.** Goldberg fixes C = 1.06 F m⁻², the γ-Al₂O₃ value of Westall and Hohl (1980).
   Carbonates use C = √I/α (Pokrovsky et al., 1999), a far stiffer double layer.
4. **Constraint.** Goldberg fits envelopes of many points over pH 3–12. This data set has five usable
   lithium points and two usable cobalt points spanning 0.08 pH units. The procedure is identical;
   what it can deliver is not, and establishing that is the result of the present work.

## References

Brady, P. V., Papenguth, H. W., & Kelly, J. W. (1999). Metal sorption to dolomite surfaces.
*Applied Geochemistry, 14*(5), 569–579.

Goldberg, S. (1985). Chemical modeling of anion competition on goethite using the constant
capacitance model. *Soil Science Society of America Journal, 49*(4), 851–856.

Pokrovsky, O. S., Schott, J., & Thomas, F. (1999). Dolomite surface speciation and reactivity in
aquatic systems. *Geochimica et Cosmochimica Acta, 63*(19–20), 3133–3143.

Westall, J. C. (1982). *FITEQL: A computer program for determination of chemical equilibrium
constants from experimental data* (Report 82-01). Oregon State University.

Westall, J. C., & Hohl, H. (1980). A comparison of electrostatic models for the oxide/solution
interface. *Advances in Colloid and Interface Science, 12*(4), 265–294.

Zachara, J. M., Cowan, C. E., & Resch, C. T. (1991). Sorption of divalent metals on calcite.
*Geochimica et Cosmochimica Acta, 55*(6), 1549–1562.